index.rst

```diff
@@ -48,14 +48,177 @@
+
+A new job would be inserted before the initial pipeline jobs in the graph that downloads the initial registry state from the cloud registry and stores it in a SQLite file.
+This SQLite file would then be provided as an input to the first job.
+The pipeline job would use that SQLite registry and a GCS-with-local-cache datastore as described in the previous section.
+On completion the output collected by the workflow execution system would be an updated SQLite registry.
+If the next pipeline job only takes files from a single input job this SQLite file would be passed directly to the next pipeline job.
+If, on the other hand, the next job gathers files from multiple jobs, the workflow generator will insert a new job before it that reads all the SQLite files and merges them into a single output for that job.
+It can optionally also sync registry information to the cloud registry and trim the SQLite file that it passes on.
+This sync and prune could be a distinct job that can be inserted after the merge job in the graph.
+At the end node (or nodes) of each workflow graph a final job will be inserted that updates the cloud registry with the final state.
+
+In this scenario pre-emption has no impact on registry state for pipeline jobs since they are running with an entirely local SQLite registry.
+Pre-emption still has to be understood in terms of the registry sync jobs and requires that the remote update of the cloud registry happens in a single transaction.
+
+This approach allows the submission system to decide whether the registry is updated multiple times within the graph or solely at the end since the registry merging jobs can be configured to either merge the inputs and pass them on complete, or merge, sync and trim.
+A trimmed registry can not be passed on to the next job if the remainder has not been synced with the cloud registry.
+
+Limited Read-Only Registry
+--------------------------
+
+When a quantum graph is constructed the graph builder knows every single input dataset and every output dataset and how they relate to each other.
+The graph builder also knows the expected URIs of all the files created by the pipelines.
+This knowledge could be used to construct a limited SQLite registry that could be constructed by the graph builder and provided to each invocation.
+BPS could for example upload this SQLite file to object store and provide a URI to each job to allow it to be retrieved.
+This static file would then be passed to every job being executed and importantly, unlike the externalized approach above, it will never need to handle merging of registry information during the execution of the workflow graph.
+
+On ``butler.put()`` the implementation would check that the relevant entry is expected but otherwise not try to do anything else.
+The datastore would also write the file to the object store and interact with registry but registry would not write anything to registry.
```

Comment:
Can this be reworded:  "but registry would not write anything to registry"  I understand that there will be no database/SQLite updates, but the wording just doesn't feel quite right.